.github/workflows/minimal-console.yml

```diff
@@ -1,68 +1,92 @@
+        type: choice
+        options:
+          - harmstone
+          - master
+          - aarch64-pe-dev
+          - aarch64-pe-zac
   pull_request:
     paths-ignore:
       - '**/*.md'
   push:
     paths-ignore:
       - '**/*.md'
 
+env:
+  default_branch: harmstone  
+
 jobs:
-  build:
-    name: Build and upload minimal console app
-    runs-on: windows-latest
+  build-binutils:
+    name: Build binutils and create object file
+    runs-on: ubuntu-latest
 
     steps:
+    - name: Install dependencies
+      run: | 
+        sudo apt-get update
+        sudo apt-get -y install build-essential binutils-for-build texinfo bison flex zlib1g-dev libgmp-dev dejagnu libmpfr-dev
+
+    - name: Download binutils-gdb source code
+      uses: actions/checkout@v3
+      with:
+        repository: Windows-on-ARM-Experiments/binutils-gdb
+        ref: ${{ inputs.branch || env.default_branch }}
+        path: binutils-gdb
+  
     - name: Checkout repository
       uses: actions/checkout@v3
-
-    - name: Create object file
-      run: .\tools\as.exe .\min-exe\con-app-aarch64.s -o con-app-aarch64.o
-
-    # # Uncomment in case there will be an issue with path overwrite by git/usr/bin/link
-    # - name: Remove /usr/bin/link
-    #   run: |
-    #     $Remove = 'C:\Program Files\Git\usr\bin'
-    #     $env:Path = ($env:Path.Split(';') | Where-Object -FilterScript {$_ -ne $Remove}) -join ';'
-    #     $env:PATH
-
-    - name: Configure build for arm64
-      uses: ilammy/msvc-dev-cmd@v1
       with:
-        arch: amd64_arm64
-
-    - name: Link object file
-      run: LINK con-app-aarch64.o kernel32.Lib /SUBSYSTEM:CONSOLE
+        path: min-con-app-source
 
-    - name: Dumpbin the executable
-      run: DUMPBIN /HEADERS .\con-app-aarch64.exe
+    - name: Configure and build
+      run: |
+        mkdir -p binutils-gdb/build
+        cd binutils-gdb/build
+        ../configure --target=aarch64-pe --prefix="$HOME/cross"
+        make -j$(nproc)
+    
+    - name: Create object file
+      run: |
+        ./binutils-gdb/build/gas/as-new ./min-con-app-source/min-exe/con-app-aarch64.s -o con-app-aarch64.o  
 
-    - name: Upload executable
+    - name: Upload object file
       uses: actions/upload-artifact@v3
       with:
-        name: min-console-exe
-        path: .\con-app-aarch64.exe
+        name: obj-file
+        path: ./con-app-aarch64.o  
 
+  build:
+    name: Build, test, create artifact
+    runs-on: [self-hosted, Windows, ARM64, GCC]
+    needs: build-binutils
 
-  test:
-    name: Run and assert the output of app run
-    runs-on: [self-hosted, Windows, ARM64]
-    needs: build
-    
     steps:
     - name: Checkout repository
       uses: actions/checkout@v3
 
-    - name: Download a single artifact
+    - name: Download object file
       uses: actions/download-artifact@v3
+      with:
+        name: obj-file
+
+    - name: Link object file
+      run: |
+        C:\Program` Files\Git\bin\bash.exe .github\workflows\scripts\link.sh
```

Comment:
call "C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Auxiliary\Build\vcvarsall.bat" arm64 could be used to get VS env
then  VCToolsInstallDir variable could be used in bash script